koku/api/settings/tags/mapping/serializers.py

```diff
@@ -37,35 +51,32 @@ def validate(self, data):
         """This function validates the options and returns the enabled tag rows."""
         children_list = data["children"]
         combined_list = [data["parent"]] + children_list
+        enabled_rows = EnabledTagKeys.objects.filter(uuid__in=combined_list, enabled=True)
+        if len(combined_list) != enabled_rows.count():
+            # Ensure that the parent & child uuids are enabled.
+            raise serializers.ValidationError("Invalid or disabled uuid provided.")
         mappings = TagMapping.objects.filter(Q(parent__uuid__in=combined_list) | Q(child__uuid__in=combined_list))
-        children_to_parent = []
-        parent_to_children = []
-        already_children = []
+        errors = defaultdict(list)
         for tag_mapping in mappings:
             if tag_mapping.parent.uuid in children_list:
-                parent_to_children.append(str(tag_mapping.parent.uuid))
+                errors["a parent cannot become a child:"].append(str(tag_mapping.parent.uuid))
             if tag_mapping.child.uuid == data["parent"]:
-                children_to_parent.append(str(tag_mapping.child.uuid))
+                errors["a child cannot become a parent:"].append(str(tag_mapping.child.uuid))
             if tag_mapping.child.uuid in children_list:
-                already_children.append(str(tag_mapping.child.uuid))
-        errors = []
-        if parent_to_children:
-            errors.append(f"a parent cannot become a child: {parent_to_children}")
-        if children_to_parent:
-            errors.append(f"a child cannot become a parent: {children_to_parent}")
-        if already_children:
-            errors.append(f"child already linked to a parent: {already_children}")
-        if errors:
-            raise serializers.ValidationError(errors)
-
-        enabled_rows = EnabledTagKeys.objects.filter(uuid__in=combined_list, enabled=True)
-        if len(combined_list) != enabled_rows.count():
-            # Ensure that the parent & child uuids are enabled.
-            raise serializers.ValidationError("Invalid or disabled uuid provided.")
+                errors["child already linked to a parent:"].append(str(tag_mapping.child.uuid))
 
+        # Checks if a child key is associated with a cost model
+        child_keys = enabled_rows.exclude(uuid=data["parent"]).values_list("key", flat=True)
+        intersecting_tags = set(child_keys) & set(self.context.keys())
+        for intersecting_tag in intersecting_tags:
+            metadata = {}
+            metadata["cost_model_id"] = self.context[intersecting_tag].get("cost_model_id")
+            metadata["child_key"] = intersecting_tag
+            errors["child is being used in a cost model:"].append(metadata)
+        if errors:
+            formatted_errors = []
+            for log_msg, log_list in errors.items():
+                format_error = [f"{log_msg} {item}" for item in log_list]
```

Comment:
I'm not sure i understand what is going on with all this formatting. Is this what we want?
```
{
    "errors": [
        {
            "detail": "child already linked to a parent: 587cbeed-8ea0-4a83-9534-d054dbbc76ed",
            "source": "non_field_errors",
            "status": 400
        },
        {
            "detail": "child already linked to a parent: c552584c-fcce-4873-a536-528105312537",
            "source": "non_field_errors",
            "status": 400
        }
    ]
}
```

Reply:
That is because I goofed it up, I switched over to using a `defaultdict(list)` to collect the errors cause I thought it was cleaner than the 3 separate lists approach. However, when I tested I forgot to add multiple errors of the same type,  I no longer need to iterate through the list. 

```
{
    "errors": [
        {
            "detail": "child already linked to a parent: ['d1376436-2726-4dad-a3c3-e45a7b9e10df', '3c21b1fd-fb98-4dce-8999-a8be4996724b']",
            "source": "non_field_errors",
            "status": 400
        },
        {
            "detail": "child is being used in a cost model: [{'cost_model_id': '581a70d3-88f5-44fa-9348-fa22a18731da', 'child_key': 'storageclass'}]",
            "source": "non_field_errors",
            "status": 400
        }
    ]
}
```